eeps/eep-0049.md

```diff
@@ -37,66 +37,73 @@ The current syntax for a `begin ... end` expression is:
+type of expressions (denoted `MatchOrReturnExprs`), only valid within a
 `begin ... end` expression:
 
     begin
-        Exprs | UnwrapExprs
+        Exprs | MatchOrReturnExprs
     end
 
-`UnwrapExprs` are defined as having the following form:
+`MatchOrReturnExprs` are defined as having the following form:
 
-    Pattern <~ Expr
+    Pattern <- Expr
 
-This definition means that `UnwrapExprs` are only allowed at the
+This definition means that `MatchOrReturnExprs` are only allowed at the
 top-level of `begin ... end` expressions.
 
-The `<~` operator takes the value return by `Expr` and inspects it.
-If the value is a tuple of the form `{ok, Val}`, it unwraps `Val` from
-the tuple, and matches it against `Pattern`.
+The `<-` operator takes the value return by `Expr` and pattern matches on
+it against `Pattern`.
 
-If the pattern matches, all variables from `Pattern` are bound in the
-local environment, and the full value `{ok, Val}` is returned by the
-`UnwrapExpr`. If the value does not match, a `{badunwrap, Val}` error
-is raised.
+If the pattern matches, all variables from `Pattern` are bound in the local
+environment, and the expression is equivalent to a successful `Pattern = Expr`
+call. If the value does not match, the `begin ... end` expression returns the
+failed expression directly.
 
-A special case exists when `Pattern` is the match-all
-variable (`_`), which on top of allowing the value to be considered a successful
-unwrapping if the returned value from `Expression` is `{ok, term()}`,
-it also considers the atom `ok` to be valid as well.
+A special case exists in which we extend `begin ... end` into the following form:
 
-If the value is a tuple of the form `{error, Reason}`, then the entire
-`begin ... end` expression is short-circuited and returns `{error,
-Reason}`. The variables that were bound in there remain bound, the
-rest are undefined.
+    begin
+        Exprs | MatchOrReturnExprs
+    else
+        Pattern -> Exprs;
+        ...
+        Pattern -> Exprs
+    end
 
-The compiler should warn about any variable that is used after the
-`begin ... end` expression that was bound in or after the first
-`UnwrapExpr` encountered within the block.
+This form exists to capture non-matching expressions in a `MatchOrReturnExprs`
+to handle failed matches rather than returning their value. In such a case, an
+unhandled failed match will raise an `else_clause` error, otherwise identical to
+a `case_clause` error.
 
-If the value returned does not match any of `{ok | error, term()}` as a
-type, a `{badunwrap, Val}` error is raised.
+This extended form is useful to properly identify and handle successful and
+unsuccessful matches within the same construct without risking to confuse
+happy and unhappy paths.
 
 Given the structure described here, the final expression may look like:
 
     begin
         Foo = bar(),
-        X <~ id({ok, 5}),
-        [H|T] <~ id({ok, [1,2,3]}),
+        {ok, X} <- id({ok, 5}),
```

Comment:
Confusing example, `Foo` is not used, and then `id/1` receives an ok tuple and a non-tupled list.